presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -179,6 +183,29 @@ public ScheduleResult schedule()
     {
         // schedule a task on every node in the distribution
         List<RemoteTask> newTasks = ImmutableList.of();
+        List<ListenableFuture<?>> blocked = new ArrayList<>();
+
+        // CTE Materialization Check
+        if (stage.requiresMaterializedCTE()) {
+            List<String> requiredCTEIds = stage.getRequiredCTEList();
+            for (String cteId : requiredCTEIds) {
+                if (!cteMaterializationTracker.hasBeenMaterialized(cteId)) {
+                    // Add CTE materialization future to the blocked list
+                    ListenableFuture<Void> materializationFuture = cteMaterializationTracker.getFutureForCTE(cteId);
+                    blocked.add(materializationFuture);
+                }
+            }
+            // If any CTE is not materialized, return a blocked ScheduleResult
+            if (!blocked.isEmpty()) {
+                return ScheduleResult.blocked(
+                        false,
+                        newTasks,
+                        whenAnyComplete(blocked),
```

Comment:
Trying to wrap my head around the logic here:

> If any CTE is not materialized, return a blocked ScheduleResult

So the logic should be, in order to return a result which isn't blocked, then the `blocked` list should be empty. This makes sense based on the `blocked.isEmpty()` condition.

However, the `blocked` future argument is `whenAnyComplete(blocked)`. Shouldn't this be something like `whenAllComplete`? Or is there some nuance to what the `blocked` argument should be to the `ScheduleResult.blocked` function? I am not very familiar with this portion of the codebase.

Reply:
I think any or all don't matter since blocked list just contains one future here

Reply:
so FixedSourcePartitionedScheduler is just for the stages reading from bucketed tables. As far as I know presto doesn't allow multiple tablescans in one stage. But this code is still resilient

Reply:
Oh yes, for hive connector, it's the bucketed tables that will expose their partitioning property into `HivePartitioningHandle` when considering colocated joins. So when left and right tables in a join have compatible bucket definition exactly on the join keys, they will be left into the same stage.

Currently, the tableScanNode for the temporary table caused by a CTE does not inherit the bucket feature of its source table, so seems that it won't stay in the same stage with other tableScanNodes. Do we have any plan to support inheriting the bucket feature of the source table in future?

Reply:
I'm still a bit confused here. A stage can have multiple CTE dependencies if they have the same bucketing, so seems like we shouldn wait for all blocked futures to complete, right? (this isn't broken now, as if we unblock when some complete, we'll just do this check again, just want to make sure i'm understanding the flow properly)